File: VelibStations_Android/app/src/main/java/com/muhammedsafiulazam/velibstations/feature/stationlist/model/StationListModel.kt
package com.muhammedsafiulazam.velibstations.feature.stationlist.model

import android.text.TextUtils
import com.muhammedsafiulazam.common.addon.AddOnType
import com.muhammedsafiulazam.common.database.IDatabaseManager
import com.muhammedsafiulazam.common.database.velib.event.VelibDatabaseEventType
import com.muhammedsafiulazam.common.event.Event
import com.muhammedsafiulazam.common.event.IEventManager
import com.muhammedsafiulazam.common.model.Location
import com.muhammedsafiulazam.common.service.IServiceManager
import com.muhammedsafiulazam.common.service.model.Error
import com.muhammedsafiulazam.common.service.velib.event.VelibServiceEventType
import com.muhammedsafiulazam.common.service.velib.model.Dataset
import com.muhammedsafiulazam.common.view.BaseModel
import com.muhammedsafiulazam.velibstations.feature.stationlist.event.StationListEventType

class StationListModel : BaseModel() {

    private lateinit var mEventManager: IEventManager
    private lateinit var mServiceManager: IServiceManager
    private lateinit var mDatabaseManager: IDatabaseManager

    private var mLocation: Location? = null

    override fun onLoad() {
        super.onLoad()

        mEventManager = getAddOn(AddOnType.EVENT_MANAGER) as IEventManager
        mServiceManager = getAddOn(AddOnType.SERVICE_MANAGER) as IServiceManager
        mDatabaseManager = getAddOn(AddOnType.DATABASE_MANAGER) as IDatabaseManager

        // Enable events.
        receiveEvents(true)
    }

    private fun responseLoadData(response: Any?, error: Error?) {
        val event = Event(StationListEventType.MODEL_RESPONSE_LOAD_DATA, response, error)
        mEventManager.send(event)
    }

    override fun onReceiveEvents(event: Event) {
        super.onReceiveEvents(event)
        if (TextUtils.equals(StationListEventType.MODEL_REQUEST_LOAD_DATA, event.type)) {
            mLocation = event.data as Location
            // Call service.
            mServiceManager.getVelibService().getData(mLocation!!.latitude, mLocation!!.longitude)

        } else if (TextUtils.equals(VelibServiceEventType.GET_DATA, event.type)) {
            if (event.error != null) {
                // Call db.
                mDatabaseManager.getVelibDatabase().getData(mLocation!!.latitude, mLocation!!.longitude)
            } else {
                // Response.
                responseLoadData(event.data, event.error)

                // Save in db.
                mDatabaseManager.getVelibDatabase().saveData(event.data as Dataset?)
            }

        } else if (TextUtils.equals(VelibDatabaseEventType.GET_DATA, event.type)) {
            // Response.
            responseLoadData(event.data, event.error)
        }
    }

    override fun onUnload() {
        // Disable events.
        receiveEvents(false)
        super.onUnload()
    }
}

File: VelibStations_iOS/VelibStations/Feature/StationList/Model/StationListModel.swift
//
//  StationListModel.swift
//  VelibStations
//
//  Created by Safi on 20/12/2019.
//

import Foundation
import CommonKit

class StationListModel : BaseModel {
    
    private var mEventManager: IEventManager? = nil
    private var mServiceManager: IServiceManager? = nil
    private var mDatabaseManager: IDatabaseManager? = nil
    private var mLocation: Location? = nil
    
    override func onLoad() {
        super.onLoad()
        
        // Addons.
        mEventManager = getAddOn(type: AddOnType().EVENT_MANAGER) as? IEventManager
        mServiceManager = getAddOn(type: AddOnType().SERVICE_MANAGER) as? IServiceManager
        mDatabaseManager = getAddOn(type: AddOnType().DATABASE_MANAGER) as? IDatabaseManager
        
        // Enable events.
        receiveEvents(receive: true)
    }
    
    private func responseLoadData(response: Any?, error: Error?) {
        let event = Event(type: StationListEventType.MODEL_RESPONSE_LOAD_DATA, data: response, error: error)
        mEventManager?.send(event: event)
    }
    
    override func onReceiveEvents(event: Event) {
        super.onReceiveEvents(event: event)
        
        if (StationListEventType.MODEL_REQUEST_LOAD_DATA == event.type) {
            mLocation = event.data as? Location
            // Call service.
            mServiceManager?.getVelibService().getData(latitude: mLocation!.latitude, longitude: mLocation!.longitude, radius: Constants().DEFAULT_RADIUS, index: Constants().DEFAULT_INDEX, count: Constants().DEFAULT_COUNT)
        } else if (VelibServiceEventType().GET_DATA == event.type) {
            if (event.error != nil) {
                // Call db.
                mDatabaseManager?.getVelibDatabase().getData(latitude: mLocation!.latitude, longitude: mLocation!.longitude, radius: Constants().DEFAULT_RADIUS)
            } else {
                // Response.
                responseLoadData(response: event.data, error: event.error)
                
                // Save in db.
                mDatabaseManager?.getVelibDatabase().saveData(dataset: event.data as? Dataset)
            }
        } else if (VelibDatabaseEventType().GET_DATA == event.type) {
            // Response.
            responseLoadData(response: event.data, error: event.error)
        }
    }
    
    override func onUnload() {
        // Disable events.
        receiveEvents(receive: false)
        super.onUnload()
    }
}
